From 74ee9d17cd3ac26f29a966f632567c55ffb868d9 Mon Sep 17 00:00:00 2001
From: Miroslav Rudisin <miero@seznam.cz>
Date: Sun, 6 Mar 2016 14:55:07 +0100
Subject: [PATCH] ASoC: mcasp: allow to change serializer cfg

---
 sound/soc/davinci/davinci-mcasp.c | 53 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 53 insertions(+)

diff --git a/sound/soc/davinci/davinci-mcasp.c b/sound/soc/davinci/davinci-mcasp.c
index 6ebbf1f..d092256 100644
--- a/sound/soc/davinci/davinci-mcasp.c
+++ b/sound/soc/davinci/davinci-mcasp.c
@@ -1219,6 +1219,58 @@ static int davinci_mcasp_hw_rule_format(struct snd_pcm_hw_params *params,
 	return snd_mask_refine(fmt, &nfmt);
 }
 
+static int davinci_mcasp_set_channel_map(struct snd_soc_dai *cpu_dai,
+		unsigned int tx_num, unsigned int *tx_slot,
+		unsigned int rx_num, unsigned int *rx_slot)
+{
+	struct davinci_mcasp *mcasp = snd_soc_dai_get_drvdata(cpu_dai);
+	unsigned int chan;
+	unsigned int slot;
+
+	/*
+	 * Notice: This sets serializer map.
+	 *         There might be multiple channels on each serializer.
+	 */
+
+	/* check for validity */
+
+	for (chan = 0; chan < tx_num; chan++) {
+		slot = tx_slot[chan];
+		if (slot >= mcasp->num_serializer) {
+			return -EINVAL;
+		}
+	}
+
+	for (chan = 0; chan < rx_num; chan++) {
+		slot = rx_slot[chan];
+		if (slot >= mcasp->num_serializer) {
+			return -EINVAL;
+		}
+	}
+
+	/* update configuration of serializers */
+
+	for (slot = 0; slot < mcasp->num_serializer; slot++) {
+		mcasp->serial_dir[slot] = INACTIVE_MODE;
+	}
+
+	for (chan = 0; chan < tx_num; chan++) {
+		slot = tx_slot[chan];
+		if (slot < mcasp->num_serializer) {
+			mcasp->serial_dir[slot] = TX_MODE;
+		}
+	}
+
+	for (chan = 0; chan < rx_num; chan++) {
+		slot = rx_slot[chan];
+		if (slot < mcasp->num_serializer) {
+			mcasp->serial_dir[slot] = RX_MODE;
+		}
+	}
+
+	return 0;
+}
+
 static int davinci_mcasp_startup(struct snd_pcm_substream *substream,
 				 struct snd_soc_dai *cpu_dai)
 {
@@ -1332,6 +1384,7 @@ static const struct snd_soc_dai_ops davinci_mcasp_dai_ops = {
 	.set_clkdiv	= davinci_mcasp_set_clkdiv,
 	.set_sysclk	= davinci_mcasp_set_sysclk,
 	.set_tdm_slot	= davinci_mcasp_set_tdm_slot,
+	.set_channel_map	= davinci_mcasp_set_channel_map,
 };
 
 static int davinci_mcasp_dai_probe(struct snd_soc_dai *dai)
-- 
2.1.4

