From b57ad8c9f29fba696e50c8f831c8be22721f559b Mon Sep 17 00:00:00 2001
From: Miroslav Rudisin <miero@seznam.cz>
Date: Sun, 6 Mar 2016 15:03:30 +0100
Subject: [PATCH] ASoC: mcasp: add support for DSD and high sampling

---
 sound/soc/davinci/davinci-mcasp.c | 70 +++++++++++++++++++++++++++++++++------
 1 file changed, 60 insertions(+), 10 deletions(-)

diff --git a/sound/soc/davinci/davinci-mcasp.c b/sound/soc/davinci/davinci-mcasp.c
index d092256..a4b9674 100644
--- a/sound/soc/davinci/davinci-mcasp.c
+++ b/sound/soc/davinci/davinci-mcasp.c
@@ -753,13 +753,13 @@ static int davinci_config_channel_size(struct davinci_mcasp *mcasp,
 }
 
 static int mcasp_common_hw_param(struct davinci_mcasp *mcasp, int stream,
-				 int period_words, int channels)
+				 int period_words, int channels, bool dsd_mode)
 {
 	struct snd_dmaengine_dai_dma_data *dma_data = &mcasp->dma_data[stream];
 	int i;
 	u8 tx_ser = 0;
 	u8 rx_ser = 0;
-	u8 slots = mcasp->tdm_slots;
+	u8 slots = dsd_mode ? 1 : mcasp->tdm_slots;
 	u8 max_active_serializers = (channels + slots - 1) / slots;
 	int active_serializers, numevt;
 	u32 reg;
@@ -862,13 +862,14 @@ static int mcasp_common_hw_param(struct davinci_mcasp *mcasp, int stream,
 }
 
 static int mcasp_i2s_hw_param(struct davinci_mcasp *mcasp, int stream,
-			      int channels)
+			      int channels, bool dsd_mode)
 {
 	int i, active_slots;
 	int total_slots;
 	int active_serializers;
 	u32 mask = 0;
 	u32 busel = 0;
+	u32 mod;
 
 	total_slots = mcasp->tdm_slots;
 
@@ -907,16 +908,24 @@ static int mcasp_i2s_hw_param(struct davinci_mcasp *mcasp, int stream,
 	if (!mcasp->dat_port)
 		busel = TXSEL;
 
+	if (dsd_mode) {
+		mask = 1;
+		busel = 0;
+		mod = 0;
+	} else {
+		mod = total_slots;
+	}
+
 	if (stream == SNDRV_PCM_STREAM_PLAYBACK) {
 		mcasp_set_reg(mcasp, DAVINCI_MCASP_TXTDM_REG, mask);
 		mcasp_set_bits(mcasp, DAVINCI_MCASP_TXFMT_REG, busel | TXORD);
 		mcasp_mod_bits(mcasp, DAVINCI_MCASP_TXFMCTL_REG,
-			       FSXMOD(total_slots), FSXMOD(0x1FF));
+			       FSXMOD(mod), FSXMOD(0x1FF));
 	} else if (stream == SNDRV_PCM_STREAM_CAPTURE) {
 		mcasp_set_reg(mcasp, DAVINCI_MCASP_RXTDM_REG, mask);
 		mcasp_set_bits(mcasp, DAVINCI_MCASP_RXFMT_REG, busel | RXORD);
 		mcasp_mod_bits(mcasp, DAVINCI_MCASP_RXFMCTL_REG,
-			       FSRMOD(total_slots), FSRMOD(0x1FF));
+			       FSRMOD(mod), FSRMOD(0x1FF));
 		/*
 		 * If McASP is set to be TX/RX synchronous and the playback is
 		 * not running already we need to configure the TX slots in
@@ -924,9 +933,12 @@ static int mcasp_i2s_hw_param(struct davinci_mcasp *mcasp, int stream,
 		 */
 		if (mcasp_is_synchronous(mcasp) && !mcasp->channels)
 			mcasp_mod_bits(mcasp, DAVINCI_MCASP_TXFMCTL_REG,
-				       FSXMOD(total_slots), FSXMOD(0x1FF));
+				       FSXMOD(mod), FSXMOD(0x1FF));
 	}
 
+	/* Disable the DIT */
+	mcasp_clr_bits(mcasp, DAVINCI_MCASP_TXDITCTL_REG, DITEN);
+
 	return 0;
 }
 
@@ -1025,16 +1037,37 @@ static int davinci_mcasp_calc_clk_div(struct davinci_mcasp *mcasp,
 	return div;
 }
 
+static int is_dsd(snd_pcm_format_t format)
+{
+	switch (format) {
+		case SNDRV_PCM_FORMAT_DSD_U8:
+		case SNDRV_PCM_FORMAT_DSD_U16_LE:
+		case SNDRV_PCM_FORMAT_DSD_U16_BE:
+		case SNDRV_PCM_FORMAT_DSD_U32_LE:
+		case SNDRV_PCM_FORMAT_DSD_U32_BE:
+			return 1;
+			break;
+
+		default:
+			return 0;
+			break;
+	}
+}
+
 static int davinci_mcasp_hw_params(struct snd_pcm_substream *substream,
 					struct snd_pcm_hw_params *params,
 					struct snd_soc_dai *cpu_dai)
 {
 	struct davinci_mcasp *mcasp = snd_soc_dai_get_drvdata(cpu_dai);
 	int word_length;
+	bool dsd_mode = is_dsd(params_format(params));
 	int channels = params_channels(params);
 	int period_size = params_period_size(params);
 	int ret;
 
+	if (dsd_mode && mcasp->op_mode == DAVINCI_MCASP_DIT_MODE)
+		return -EINVAL;
+
 	/*
 	 * If mcasp is BCLK master, and a BCLK divider was not provided by
 	 * the machine driver, we need to calculate the ratio.
@@ -1058,7 +1091,7 @@ static int davinci_mcasp_hw_params(struct snd_pcm_substream *substream,
 	}
 
 	ret = mcasp_common_hw_param(mcasp, substream->stream,
-				    period_size * channels, channels);
+				    period_size * channels, channels, dsd_mode);
 	if (ret)
 		return ret;
 
@@ -1066,7 +1099,7 @@ static int davinci_mcasp_hw_params(struct snd_pcm_substream *substream,
 		ret = mcasp_dit_hw_param(mcasp, params_rate(params));
 	else
 		ret = mcasp_i2s_hw_param(mcasp, substream->stream,
-					 channels);
+					 channels, dsd_mode);
 
 	if (ret)
 		return ret;
@@ -1097,6 +1130,20 @@ static int davinci_mcasp_hw_params(struct snd_pcm_substream *substream,
 		word_length = 32;
 		break;
 
+	case SNDRV_PCM_FORMAT_DSD_U8:
+		word_length = 8;
+		break;
+
+	case SNDRV_PCM_FORMAT_DSD_U16_LE:
+	case SNDRV_PCM_FORMAT_DSD_U16_BE:
+		word_length = 16;
+		break;
+
+	case SNDRV_PCM_FORMAT_DSD_U32_LE:
+	case SNDRV_PCM_FORMAT_DSD_U32_BE:
+		word_length = 32;
+		break;
+
 	default:
 		printk(KERN_WARNING "davinci-mcasp: unsupported PCM format");
 		return -EINVAL;
@@ -1465,7 +1512,7 @@ static int davinci_mcasp_resume(struct snd_soc_dai *dai)
 #define davinci_mcasp_resume NULL
 #endif
 
-#define DAVINCI_MCASP_RATES	SNDRV_PCM_RATE_8000_192000
+#define DAVINCI_MCASP_RATES	(SNDRV_PCM_RATE_8000_768000 | SNDRV_PCM_RATE_CONTINUOUS)
 
 #define DAVINCI_MCASP_PCM_FMTS (SNDRV_PCM_FMTBIT_S8 | \
 				SNDRV_PCM_FMTBIT_U8 | \
@@ -1476,7 +1523,10 @@ static int davinci_mcasp_resume(struct snd_soc_dai *dai)
 				SNDRV_PCM_FMTBIT_S24_3LE | \
 				SNDRV_PCM_FMTBIT_U24_3LE | \
 				SNDRV_PCM_FMTBIT_S32_LE | \
-				SNDRV_PCM_FMTBIT_U32_LE)
+				SNDRV_PCM_FMTBIT_U32_LE | \
+				SNDRV_PCM_FMTBIT_DSD_U8 | \
+				SNDRV_PCM_FMTBIT_DSD_U16_LE | \
+				SNDRV_PCM_FMTBIT_DSD_U32_LE)
 
 static struct snd_soc_dai_driver davinci_mcasp_dai[] = {
 	{
-- 
2.1.4

