From 0564cd542cb38c347d10a4e8521ebf68bdfbef23 Mon Sep 17 00:00:00 2001
From: Miroslav Rudisin <miero@seznam.cz>
Date: Sun, 24 Apr 2016 16:13:26 +0200
Subject: [PATCH] ASoC: mcasp: disable unnecessary pins for DIT & DSD playback

---
 sound/soc/davinci/davinci-mcasp.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/sound/soc/davinci/davinci-mcasp.c b/sound/soc/davinci/davinci-mcasp.c
index c6f0629..ef0f836 100644
--- a/sound/soc/davinci/davinci-mcasp.c
+++ b/sound/soc/davinci/davinci-mcasp.c
@@ -787,12 +787,22 @@ static int mcasp_common_hw_param(struct davinci_mcasp *mcasp, int stream,
 	u8 max_active_serializers = (channels + slots - 1) / slots;
 	int active_serializers, numevt;
 	u32 reg;
+	u32 disable_pins;
+
 	/* Default configuration */
 	if (mcasp->version < MCASP_VERSION_3)
 		mcasp_set_bits(mcasp, DAVINCI_MCASP_PWREMUMGT_REG, MCASP_SOFT);
 
-	/* All PINS as McASP */
-	mcasp_set_reg(mcasp, DAVINCI_MCASP_PFUNC_REG, 0x00000000);
+	if (mcasp->op_mode == DAVINCI_MCASP_DIT_MODE) {
+		disable_pins = AFSX | ACLKX;
+	} else if (dsd_mode) {
+		disable_pins = AFSX;
+	} else {
+		disable_pins = 0;
+	}
+	mcasp_set_reg(mcasp, DAVINCI_MCASP_PFUNC_REG, disable_pins);
+	mcasp_clr_bits(mcasp, DAVINCI_MCASP_PDOUT_REG, disable_pins);
+	mcasp_set_bits(mcasp, DAVINCI_MCASP_PDIR_REG, disable_pins);
 
 	if (stream == SNDRV_PCM_STREAM_PLAYBACK) {
 		mcasp_set_reg(mcasp, DAVINCI_MCASP_TXSTAT_REG, 0xFFFFFFFF);
-- 
2.1.4

